File: packages/go/pricing.go
package aikit

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
)

type curatedModel struct {
	ID            string            `json:"id"`
	DisplayName   string            `json:"displayName"`
	Provider      Provider          `json:"provider"`
	Family        string            `json:"family"`
	Capabilities  ModelCapabilities `json:"capabilities"`
	ContextWindow int               `json:"contextWindow"`
	TokenPrices   *TokenPrices      `json:"tokenPrices"`
	Deprecated    bool              `json:"deprecated"`
	InPreview     bool              `json:"inPreview"`
}

var curatedOnce sync.Once
var curatedModels []curatedModel

func loadCuratedModels() []curatedModel {
	curatedOnce.Do(func() {
		path := curatedModelsPath()
		if path == "" {
			curatedModels = []curatedModel{}
			return
		}
		raw, err := os.ReadFile(path)
		if err != nil {
			curatedModels = []curatedModel{}
			return
		}
		if err := json.Unmarshal(raw, &curatedModels); err != nil {
			curatedModels = []curatedModel{}
			return
		}
	})
	return curatedModels
}

func curatedModelsPath() string {
	_, filename, _, ok := runtime.Caller(0)
	if !ok {
		return ""
	}
	return filepath.Join(filepath.Dir(filename), "..", "..", "models", "curated_models.json")
}

func normalizeModelID(provider Provider, modelID string) string {
	prefix := string(provider) + "/"
	if strings.HasPrefix(modelID, prefix) {
		return strings.TrimPrefix(modelID, prefix)
	}
	return modelID
}

func findCuratedModel(provider Provider, modelID string) *curatedModel {
	normalized := normalizeModelID(provider, modelID)
	var best *curatedModel
	for i := range loadCuratedModels() {
		model := &curatedModels[i]
		if model.Provider != provider {
			continue
		}
		if model.ID == normalized {
			return model
		}
		if strings.HasPrefix(normalized, model.ID) {
			if best == nil || len(model.ID) > len(best.ID) {
				best = model
			}
		}
	}
	return best
}

func applyCuratedMetadata(model ModelMetadata) ModelMetadata {
	curated := findCuratedModel(model.Provider, model.ID)
	if curated == nil {
		return model
	}
	if curated.DisplayName != "" {
		model.DisplayName = curated.DisplayName
	}
	if curated.Family != "" {
		model.Family = curated.Family
	}
	model.Capabilities = curated.Capabilities
	if curated.ContextWindow > 0 {
		model.ContextWindow = curated.ContextWindow
	}
	if curated.TokenPrices != nil {
		model.TokenPrices = curated.TokenPrices
	}
	model.Deprecated = curated.Deprecated
	model.InPreview = curated.InPreview
	return model
}

func lookupTokenPrices(provider Provider, modelID string) *TokenPrices {
	curated := findCuratedModel(provider, modelID)
	if curated == nil {
		return nil
	}
	return curated.TokenPrices
}

func estimateCost(provider Provider, modelID string, usage *Usage) *CostBreakdown {
	if usage == nil {
		return nil
	}
	pricing := lookupTokenPrices(provider, modelID)
	if pricing == nil {
		return nil
	}
	if pricing.Input == 0 && pricing.Output == 0 {
		return nil
	}
	inputCost := float64(usage.InputTokens) * pricing.Input / 1_000_000
	outputCost := float64(usage.OutputTokens) * pricing.Output / 1_000_000
	return &CostBreakdown{
		InputCostUSD:    roundUsd(inputCost),
		OutputCostUSD:   roundUsd(outputCost),
		TotalCostUSD:    roundUsd(inputCost + outputCost),
		PricingPerMillion: pricing,
	}
}

func roundUsd(value float64) float64 {
	return math.Round(value*1_000_000) / 1_000_000
}

File: packages/node/src/core/pricing.ts
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import type {
  CostBreakdown,
  ModelCapabilities,
  ModelMetadata,
  Provider,
  TokenPrices,
  Usage,
} from "./types.js";

interface CuratedModel {
  id: string;
  provider: Provider;
  displayName?: string;
  family?: string;
  capabilities?: ModelCapabilities;
  contextWindow?: number;
  tokenPrices?: TokenPrices;
  videoPrices?: Record<string, number>;
  deprecated?: boolean;
  inPreview?: boolean;
}

let cachedCuratedModels: CuratedModel[] | null = null;

function modelsRoot(): string {
  return fileURLToPath(new URL("../../../../models", import.meta.url));
}

function loadCuratedModels(): CuratedModel[] {
  if (cachedCuratedModels) {
    return cachedCuratedModels;
  }
  const collected: CuratedModel[] = [];
  try {
    const entries = fs.readdirSync(modelsRoot(), { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) {
        continue;
      }
      const provider = entry.name;
      const filePath = path.join(modelsRoot(), provider, "scraped_models.json");
      if (!fs.existsSync(filePath)) {
        continue;
      }
      const raw = fs.readFileSync(filePath, "utf-8");
      const parsed = JSON.parse(raw);
      if (!Array.isArray(parsed)) {
        continue;
      }
      for (const item of parsed) {
        if (!item || typeof item !== "object") {
          continue;
        }
        if (!("provider" in item)) {
          (item as CuratedModel).provider = provider as Provider;
        }
        collected.push(item as CuratedModel);
      }
    }
  } catch {
    // ignore and return empty list
  }
  cachedCuratedModels = collected;
  return cachedCuratedModels;
}

function normalizeModelId(provider: Provider, modelId: string): string {
  const prefix = `${provider}/`;
  if (modelId.startsWith(prefix)) {
    return modelId.slice(prefix.length);
  }
  return modelId;
}

function findCuratedModel(
  provider: Provider,
  modelId: string,
): CuratedModel | undefined {
  const normalized = normalizeModelId(provider, modelId);
  let best: CuratedModel | undefined;
  for (const model of loadCuratedModels()) {
    if (model.provider !== provider) {
      continue;
    }
    if (model.id === normalized) {
      return model;
    }
    if (normalized.startsWith(model.id)) {
      if (!best || model.id.length > best.id.length) {
        best = model;
      }
    }
  }
  return best;
}

export function applyCuratedMetadata(model: ModelMetadata): ModelMetadata {
  const curated = findCuratedModel(model.provider, model.id);
  if (!curated) {
    return model;
  }
  return {
    ...model,
    displayName: curated.displayName ?? model.displayName,
    family: curated.family ?? model.family,
    capabilities: curated.capabilities ?? model.capabilities,
    contextWindow: curated.contextWindow ?? model.contextWindow,
    tokenPrices: curated.tokenPrices ?? model.tokenPrices,
    videoPrices: curated.videoPrices ?? model.videoPrices,
    deprecated: curated.deprecated ?? model.deprecated,
    inPreview: curated.inPreview ?? model.inPreview,
  };
}

export function lookupTokenPrices(
  provider: Provider,
  modelId: string,
): TokenPrices | undefined {
  return findCuratedModel(provider, modelId)?.tokenPrices;
}

export function estimateCost(
  provider: Provider,
  modelId: string,
  usage?: Usage,
): CostBreakdown | undefined {
  if (!usage) {
    return undefined;
  }
  const pricing = lookupTokenPrices(provider, modelId);
  if (!pricing) {
    return undefined;
  }
  const inputRate = pricing.input ?? 0;
  const outputRate = pricing.output ?? 0;
  if (!inputRate && !outputRate) {
    return undefined;
  }
  if (usage.inputTokens === undefined && usage.outputTokens === undefined) {
    return undefined;
  }
  const inputTokens = usage.inputTokens ?? 0;
  const outputTokens = usage.outputTokens ?? 0;
  const inputCost = (inputTokens * inputRate) / 1_000_000;
  const outputCost = (outputTokens * outputRate) / 1_000_000;
  return {
    input_cost_usd: roundUsd(inputCost),
    output_cost_usd: roundUsd(outputCost),
    total_cost_usd: roundUsd(inputCost + outputCost),
    pricing_per_million: pricing,
  };
}

function roundUsd(value: number): number {
  return Math.round(value * 1_000_000) / 1_000_000;
}
